File: dotnet/Base/OpenStack/Platform_System.cs
using Microsoft.Extensions.FileSystemGlobbing;
using OpenStack.Sfx;
using System;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;
using System.Linq;

namespace OpenStack;

/// <summary>
/// SystemAudioBuilder
/// </summary>
public class SystemAudioBuilder : AudioBuilderBase<object> {
    public override object CreateAudio(object path) => throw new NotImplementedException();
    public override void DeleteAudio(object audio) => throw new NotImplementedException();
}

#region OpenSfx

/// <summary>
/// SystemSfx
/// </summary>
public class SystemSfx(ISource source) : IOpenSfx<object> {
    readonly ISource _source = source;
    readonly AudioManager<object> _audioManager = new(source, new SystemAudioBuilder());

    public ISource Source => _source;
    public AudioManager<object> AudioManager => _audioManager;
    public object CreateAudio(object path) => _audioManager.CreateAudio(path).aud;
}

#endregion

#region FileSystem

/// <summary>
/// AggregateFileSystem
/// </summary>
public class AggregateFileSystem(IFileSystem[] aggreate) : IFileSystem {
    readonly IFileSystem[] aggreate = aggreate;

    public IEnumerable<string> Glob(string path, string searchPattern) => aggreate.SelectMany(x => x.Glob(path, searchPattern));
    public bool FileExists(string path) => aggreate.Any(x => x.FileExists(path));
    public (string path, long length) FileInfo(string path) => aggreate.Select(x => x.FileInfo(path)).FirstOrDefault(x => x.path != null);
    public BinaryReader OpenReader(string path) => aggreate.Select(x => x.OpenReader(path)).FirstOrDefault(x => x != null);
    public BinaryWriter OpenWriter(string path) => aggreate.Select(x => x.OpenWriter(path)).FirstOrDefault(x => x != null);
}

/// <summary>
/// HostFileSystem
/// </summary>
public class HostFileSystem : IFileSystem {
    public HostFileSystem(Uri uri) {
        if (uri == null) throw new ArgumentNullException(nameof(uri));
        var pathOrPattern = uri.LocalPath;
        var searchPattern = Path.GetFileName(pathOrPattern);
        //var path = Path.GetDirectoryName(pathOrPattern);
        // file
        if (!string.IsNullOrEmpty(searchPattern)) throw new ArgumentOutOfRangeException(nameof(pathOrPattern), pathOrPattern); //: Web single file access to supported.

        //options = PakOption.Stream;
        //searchPattern = Path.GetFileName(path);
        //path = Path.GetDirectoryName(path);
        //if (path.Contains('*')) throw new NotSupportedException("Web wildcard folder access");
        //host = new UriBuilder(uri) { Path = $"{path}/", Fragment = null }.Uri;
        //if (searchPattern.Contains('*'))
        //{
        //    var set = new HttpHost(host).GetSetAsync().Result ?? throw new NotSupportedException(".set not found. Web wildcard access");
        //    var pattern = $"^{Regex.Escape(searchPattern.Replace('*', '%')).Replace("_", ".").Replace("%", ".*")}$";
        //    return set.Where(x => Regex.IsMatch(x, pattern)).ToArray();
        //}
        //return new[] { searchPattern };
    }
    public IEnumerable<string> Glob(string path, string searchPattern) {
        var matcher = new Matcher();
        matcher.AddIncludePatterns([searchPattern]);
        return matcher.GetResultsInFullPath(searchPattern).ToList();
    }
    public bool FileExists(string path) => File.Exists(path);
    public (string path, long length) FileInfo(string path) => File.Exists(path) ? (path, 0) : (null, 0);
    public BinaryReader OpenReader(string path) => null;
    public BinaryWriter OpenWriter(string path) => null;
}

/// <summary>
/// IsoFileSystem
/// </summary>
public class IsoFileSystem(string root, string path) : IFileSystem {
    //readonly ZipArchive Pak = ZipFile.Open(root, ZipArchiveMode.Read);
    //readonly string Root = path;

    //public IEnumerable<string> Glob(string path, string searchPattern) {
    //    var matcher = PlatformX.CreateMatcher(searchPattern);
    //    return Pak.Entries.Where(x => matcher(x.Name)).Select(x => x.Name);
    //}
    //public bool FileExists(string path) => Pak.GetEntry(path) != null;
    //public (string path, long length) FileInfo(string path) { var x = Pak.GetEntry(path); return x != null ? (x.Name, x.Length) : (null, 0); }
    //public BinaryReader OpenReader(string path) => new(Pak.GetEntry(path).Open());
    //public BinaryWriter OpenWriter(string path) => throw new NotSupportedException();

    public bool FileExists(string path) => throw new NotImplementedException();
    public (string path, long length) FileInfo(string path) => throw new NotImplementedException();
    public IEnumerable<string> Glob(string path, string searchPattern) => throw new NotImplementedException();
    public BinaryReader OpenReader(string path) => throw new NotImplementedException();
    public BinaryWriter OpenWriter(string path) => throw new NotImplementedException();

    public IFileSystem Advance(ref string root) => throw new NotImplementedException();
}

public class SingleFileSystem(Stream stream) : IFileSystem {
    readonly Stream Stream = stream;

    public IEnumerable<string> Glob(string path, string searchPattern) => throw new NotImplementedException();
    public bool FileExists(string path) => throw new NotImplementedException();
    public (string path, long length) FileInfo(string path) => throw new NotImplementedException();
    public BinaryReader OpenReader(string path) => throw new NotImplementedException();
    public BinaryWriter OpenWriter(string path) => throw new NotImplementedException();
}

/// <summary>
/// StandardFileSystem
/// </summary>
public class StandardFileSystem(string root) : IFileSystem {
    readonly string Root = root; readonly int Skip = root.Length + 1;

    public IEnumerable<string> Glob(string path, string searchPattern) {
        var matcher = new Matcher();
        matcher.AddIncludePatterns([string.IsNullOrEmpty(searchPattern) ? "**/*" : searchPattern]);
        return [.. matcher.GetResultsInFullPath(Path.Combine(Root, path)).Select(x => x[Skip..])];
    }
    public bool FileExists(string path) => File.Exists(Path.Combine(Root, path));
    public (string path, long length) FileInfo(string path) => File.Exists(path = Path.Combine(Root, path)) ? (path[Skip..], new FileInfo(Path.Combine(Root, path)).Length) : (null, 0);
    public BinaryReader OpenReader(string path) => new(File.Open(Path.Combine(Root, path), FileMode.Open, FileAccess.Read, FileShare.Read));
    public BinaryWriter OpenWriter(string path) => new(File.Open(Path.Combine(Root, path), FileMode.Open, FileAccess.Write, FileShare.Write));
}

/// <summary>
/// VirtualFileSystem
/// </summary>
public class VirtualFileSystem(Dictionary<string, byte[]> virtuals) : IFileSystem {
    readonly Dictionary<string, byte[]> Virtuals = virtuals;

    public IEnumerable<string> Glob(string path, string searchPattern) {
        var matcher = PlatformX.CreateMatcher(searchPattern);
        return Virtuals.Keys.Where(matcher);
    }
    public bool FileExists(string path) => Virtuals.ContainsKey(path);
    public (string path, long length) FileInfo(string path) => Virtuals.TryGetValue(path, out var z) ? (path, z != null ? z.Length : 0) : (null, 0);
    public BinaryReader OpenReader(string path) => Virtuals.TryGetValue(path, out var z) ? new BinaryReader(z != null ? new MemoryStream(z) : new MemoryStream()) : null;
    public BinaryWriter OpenWriter(string path) => throw new NotSupportedException();
}

/// <summary>
/// ZipFileSystem
/// </summary>
public class ZipFileSystem(string root, string path) : IFileSystem {
    readonly ZipArchive Zip = ZipFile.Open(root, ZipArchiveMode.Read);
    readonly string Root = string.IsNullOrEmpty(path) ? string.Empty : $"{path}{Path.AltDirectorySeparatorChar}";

    public IEnumerable<string> Glob(string path, string searchPattern) {
        var root = Path.Combine(Root, path); var skip = root.Length;
        var matcher = PlatformX.CreateMatcher(searchPattern);
        return [.. Zip.Entries.Where(x =>
        {
            var fn = x.FullName;
            return fn.Length > skip && fn.StartsWith(root) && matcher(fn[skip..]);
        }).Select(x => x.FullName[skip..])];
    }
    public bool FileExists(string path) => Zip.GetEntry(Path.Combine(Root, path)) != null;
    public (string path, long length) FileInfo(string path) { var x = Zip.GetEntry(Path.Combine(Root, path)); return x != null ? (x.Name, x.Length) : (null, 0); }
    public BinaryReader OpenReader(string path) => new(Zip.GetEntry(Path.Combine(Root, path)).Open());
    public BinaryWriter OpenWriter(string path) => throw new NotSupportedException();

    public IFileSystem Advance(ref string root) {
        if (Zip.Entries.Count != 1) return this;
        var entry = Zip.Entries[0];
        root = entry.Name;
        return new SingleFileSystem(entry.Open());
    }
}

#endregion

File: Python/openstk/platforms/platform_system.py

from __future__ import annotations
import os, io, pathlib
from zipfile import ZipFile
from openstk import ISource, BinaryReader
from openstk.sfx import IOpenSfx2, AudioBuilderBase, AudioManager
from openstk.platforms import IFileSystem, PlatformX

#region OpenSfx

# SystemAudioBuilder
class SystemAudioBuilder(AudioBuilderBase):
    def createAudio(self, path: object) -> object: raise NotImplementedError()
    def deleteAudio(self, audio: object) -> None: raise NotImplementedError()

# SystemSfx
class SystemSfx(IOpenSfx2):
    source: ISource
    audioManager: AudioManager
    def __init__(self, source: ISource):
        self.source = source
        self.audioManager = AudioManager(source, SystemAudioBuilder())
    def createAudio(self, path: object) -> int: return self.audioManager.createAudio(path)[0]

#endregion

#region FileSystem

# tag::AggregateFileSystem[]
class AggregateFileSystem(IFileSystem):
    def __init__(self, aggreate: list[IFileSystem]): self.aggreate = aggreate
    def glob(self, path: str, searchPattern: str) -> list[str]: return [y for z in [x.glob(path, searchPattern) for x in self.aggreate] for y in z]
    def fileExists(self, path: str) -> bool: return any(x.fileExists(path) for x in self.aggreate)
    def fileInfo(self, path: str) -> (str, int): return min(x.fileInfo(path) for x in self.aggreate if x[0])
    def openReader(self, path: str, mode: str = 'rb') -> BinaryReader: return min(x.openReader(path) for x in self.aggreate if x)
    # def glob(self, path: str, searchPattern: str) -> list[str]:
    #     matcher = PlatformX.createMatcher(searchPattern)
    #     return [x for x in self.virtuals.keys() if matcher(x)] + self.base.glob(path, searchPattern)
    # def fileExists(self, path: str) -> bool: return path in self.virtuals or self.base.fileExists(path)
    # def fileInfo(self, path: str) -> (str, int): return (path, len(self.virtuals[path]) if (self.virtuals[path]) else 0) if path in self.virtuals else self.base.fileInfo(path)
    # def openReader(self, path: str, mode: str = 'rb') -> BinaryReader: return BinaryReader(self.virtuals[path] or io.BytesIO()) if path in self.virtuals else self.base.openReader(path)
# end::StandardFileSystem[]

# tag::HostFileSystem[]
class HostFileSystem(IFileSystem):
    def __init__(self, uri: str): self.uri = uri
    def glob(self, path: str, searchPattern: str) -> list[str]: raise NotImplementedError()
    def fileExists(self, path: str) -> bool: raise NotImplementedError()
    def fileInfo(self, path: str) -> (str, int): raise NotImplementedError()
    def openReader(self, path: str, mode: str = 'rb') -> BinaryReader: raise NotImplementedError()
# end::HostFileSystem[]

# tag::IsoFileSystem[]
class IsoFileSystem(IFileSystem):
    def __init__(self, root: str, path: str): self.pak = ZipFile(root); self.root = '' if not path else f'{path}/'
    def glob(self, path: str, searchPattern: str) -> list[str]: root = os.path.join(self.root, path); skip = len(root); return []
    def fileExists(self, path: str) -> bool: return self.pak.read(path) != None
    def fileInfo(self, path: str) -> (str, int): x = self.pak.read(path); return (x.name, x.length) if x else (None, 0)
    def openReader(self, path: str, mode: str = 'rb') -> BinaryReader: return BinaryReader(self.pak.read(path))
# end::IsoFileSystem[]

# tag::StandardFileSystem[]
class StandardFileSystem(IFileSystem):
    def __init__(self, root: str): self.root = root; self.skip = len(root) + 1
    def glob(self, path: str, searchPattern: str) -> list[str]:
        g = pathlib.Path(os.path.join(self.root, path)).glob(searchPattern if searchPattern else '**/*')
        return [str(x)[self.skip:] for x in g if x.is_file()]
    def fileExists(self, path: str) -> bool: return os.path.exists(os.path.join(self.root, path))
    def fileInfo(self, path: str) -> (str, int): return (path, os.stat(path).st_size) if os.path.exists(path := os.path.join(self.root, path)) else (None, 0)
    def openReader(self, path: str, mode: str = 'rb') -> BinaryReader: return BinaryReader(open(os.path.join(self.root, path), mode))
# end::StandardFileSystem[]

# tag::VirtualFileSystem[]
class VirtualFileSystem(IFileSystem):
    def __init__(self, virtuals: dict[str, object]): self.virtuals = virtuals
    def glob(self, path: str, searchPattern: str) -> list[str]:
        matcher = PlatformX.createMatcher(searchPattern)
        return [x for x in self.virtuals.keys() if matcher(x)]
    def fileExists(self, path: str) -> bool: return path in self.virtuals
    def fileInfo(self, path: str) -> (str, int): return (path, len(self.virtuals[path]) if (self.virtuals[path]) else 0) if path in self.virtuals else (None, 0)
    def openReader(self, path: str, mode: str = 'rb') -> BinaryReader: return BinaryReader(self.virtuals[path] or io.BytesIO()) if path in self.virtuals else None
# end::StandardFileSystem[]

# tag::ZipFileSystem[]
class ZipFileSystem(IFileSystem):
    def __init__(self, root: str, path: str): self.zip = ZipFile(root); self.root = '' if not path else f'{path}/'
    def glob(self, path: str, searchPattern: str) -> list[str]: root = os.path.join(self.root, path); skip = len(root); return []
    def fileExists(self, path: str) -> bool: return self.zip.read(os.path.join(self.root, path)) != None
    def fileInfo(self, path: str) -> (str, int): x = self.zip.read(os.path.join(self.root, path)); return (x.name, x.length) if x else (None, 0)
    def openReader(self, path: str, mode: str = 'rb') -> BinaryReader: return BinaryReader(self.zip.read(os.path.join(self.root, path)))

#endregion